sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -0,0 +1,156 @@
+from .types import X_DTYPE, X_BINNED_DTYPE
+
+
+def _find_binning_thresholds(data, max_bins, subsample, random_state):
+    """Extract feature-wise quantiles from numerical data.
+
+    Parameters
+    ----------
+    data: array-like
+        The data to bin
+    max_bins : int
+        The maximum number of bins to use. If for a given feature the number of
+        unique values is less than ``max_bins``, then those unique values
+        will be used to compute the bin thresholds, instead of the quantiles.
+    subsample : int or None
+        If ``n_samples > subsample``, then ``sub_samples`` samples will be
+        randomly choosen to compute the quantiles. If ``None``, the whole data
+        is used.
+    random_state: int or numpy.random.RandomState or None
+        Pseudo-random number generator to control the random sub-sampling.
+        See :term:`random_state`.
+
+    Return
+    ------
+    binning_thresholds: tuple of arrays
+        For each feature, stores the increasing numeric values that can
+        be used to separate the bins. len(binning_thresholds) == n_features.
+    """
+    if not (2 <= max_bins <= 256):
+        raise ValueError('max_bins={} should be no smaller than 2 '
+                         'and no larger than 256.'.format(max_bins))
+    rng = check_random_state(random_state)
+    if subsample is not None and data.shape[0] > subsample:
+        subset = rng.choice(np.arange(data.shape[0]), subsample)
+        data = data[subset]
+
+    percentiles = np.linspace(0, 100, num=max_bins + 1)
+    end = percentiles.shape[0]  # no negative indexing!
+    percentiles = percentiles[1:end - 1]
+    binning_thresholds = []
+    for f_idx in range(data.shape[1]):
+        col_data = np.ascontiguousarray(data[:, f_idx], dtype=X_DTYPE)
+        distinct_values = np.unique(col_data)
+        if len(distinct_values) <= max_bins:
+            end = distinct_values.shape[0]  # no negative indexing!
+            midpoints = (distinct_values[:end - 1] + distinct_values[1:])
+            midpoints *= .5
+        else:
+            # We sort again the data in this case. We could compute
+            # approximate midpoint percentiles using the output of
+            # np.unique(col_data, return_counts) instead but this is more
+            # work and the performance benefit will be limited because we
+            # work on a fixed-size subsample of the full data.
+            midpoints = np.percentile(col_data, percentiles,
+                                      interpolation='midpoint').astype(X_DTYPE)
+        binning_thresholds.append(midpoints)
+    return binning_thresholds
+
+
+class _BinMapper(BaseEstimator, TransformerMixin):
+    """Transformer that maps a dataset into integer-valued bins.
+
+    The bins are created in a feature-wise fashion, using quantiles so that
+    each bins contains approximately the same number of samples.
+
+    Large datasets are subsampled, but the feature-wise quantiles should
```

Comment:
It might be good to mention that the quantiles are computed on the subset. I was kinda of missing it.

For large datasets, quantiles are computed on a subset of the data to speed-up the binning. However, the feature-wise quantiles should remain stable.